File: day13/puzzle.ts
interface Offset {
  x: number;
  y: number;
}

interface ClawMachine {
  buttonA: Offset;
  buttonB: Offset;
  prize: Offset;
}

interface Pushes {
  buttonA: number;
  buttonB: number;
}

type Input = ClawMachine[];

function getPushes(machine: ClawMachine): Pushes | null {
  const buttonA = (machine.prize.y * machine.buttonB.x - machine.prize.x * machine.buttonB.y) / (machine.buttonA.y * machine.buttonB.x - machine.buttonB.y * machine.buttonA.x);
  if (Math.floor(buttonA) !== buttonA) {
    return null;
  }
  const buttonB = (machine.prize.x - buttonA * machine.buttonA.x) / machine.buttonB.x;
  if (Math.floor(buttonB) !== buttonB) {
    return null;
  }
  return { buttonA, buttonB };
}

function getTokens(pushes: Pushes[]) {
  return pushes.reduce((tokens, push) => tokens + push.buttonA * 3 + push.buttonB, 0);
}

export function parseInput(rawInput: string): Input {
  const buttonRe = /X\+(\d+), Y\+(\d+)/;
  const prizeRe = /X=(\d+), Y=(\d+)/;
  const machines: ClawMachine[] = [];
  for (const section of rawInput.split("\n\n")) {
    const lines = section.split("\n");
    const buttonA = buttonRe.exec(lines[0])!;
    const buttonB = buttonRe.exec(lines[1])!;
    const prize = prizeRe.exec(lines[2])!;
    machines.push({
      buttonA: { x: +buttonA[1], y: +buttonA[2] },
      buttonB: { x: +buttonB[1], y: +buttonB[2] },
      prize: { x: +prize[1], y: +prize[2] },
    });
  }
  return machines;
}

export function runPartOne(input: Input) {
  const allPushes = input.map(getPushes).filter(push => push && push.buttonA <= 100 && push.buttonB <= 100);
  return getTokens(allPushes as Pushes[]);
}

export function runPartTwo(input: Input) {
  for (const machine of input) {
    machine.prize.x += 10000000000000;
    machine.prize.y += 10000000000000;
  }
  const allPushes = input.map(getPushes).filter(Boolean);
  return getTokens(allPushes as Pushes[]);
}

File: day13-fs/Puzzle.fs
module Puzzle

open System.Text.RegularExpressions

type Offset = { X: int64; Y: int64 }
type ClawMachine = { ButtonA: Offset; ButtonB: Offset; Prize: Offset }
type private Pushes = { ButtonA: int64; ButtonB: int64 }

let private buttonRegex = Regex(@"X\+(\d+), Y\+(\d+)")
let private prizeRegex = Regex(@"X=(\d+), Y=(\d+)")

let private createMachine (lines: string[]): ClawMachine =
    let buttonA = buttonRegex.Match(lines[0])
    let buttonB = buttonRegex.Match(lines[1])
    let prize = prizeRegex.Match(lines[2])
    { ButtonA = { X = int64 buttonA.Groups[1].Value; Y = int64 buttonA.Groups[2].Value }
      ButtonB = { X = int64 buttonB.Groups[1].Value; Y = int64 buttonB.Groups[2].Value }
      Prize = { X = int64 prize.Groups[1].Value; Y = int64 prize.Groups[2].Value } }

let private getPushes (machine: ClawMachine) =
    let buttonA = float (machine.Prize.Y * machine.ButtonB.X - machine.Prize.X * machine.ButtonB.Y) / float (machine.ButtonA.Y * machine.ButtonB.X - machine.ButtonB.Y * machine.ButtonA.X)
    if buttonA > floor buttonA then
        None
    else
        let buttonB = float (machine.Prize.X - int64 buttonA * machine.ButtonA.X) / float machine.ButtonB.X
        if buttonB > floor buttonB then
            None
        else
            Some({ ButtonA = int64 buttonA; ButtonB = int64 buttonB })

let private getTokens (pushes: Pushes list) = pushes |> List.sumBy (fun push -> push.ButtonA * 3L + push.ButtonB)

let parseInput (rawInput: string) =
    let rec loop (sections: string list) (machines: ClawMachine list) =
        match sections with
        | [] -> machines
        | head :: tail -> loop tail ((head.Split('\n') |> createMachine) :: machines)

    loop (rawInput.Split("\n\n") |> Array.toList) []

let runPartOne (input: ClawMachine list) =
    input
    |> List.map getPushes
    |> List.choose (function
        | Some(push) when push.ButtonA <= 100 && push.ButtonB <= 100 -> Some(push)
        | _ -> None)
    |> getTokens

let runPartTwo (input: ClawMachine list) =
    input
    |> List.map (fun machine -> getPushes { machine with Prize = { X = machine.Prize.X + 10000000000000L; Y = machine.Prize.Y + 10000000000000L } })
    |> List.choose id
    |> getTokens
